### PR TITLE
remove read client/handler and write handler when corresponding fabric is removed

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
https://github.com/project-chip/connectedhomeip/issues/21084

#### Change overview
remove read client, read handler and write handler when corresponding fabric is removed

#### Testing
add unit testing to validate if read client and handler can be deleted when its fabric is removed.
add unit testing to validate if write handler can be deleted when its fabric is removed.

### CODE REVIEW

File: src/app/InteractionModelEngine.cpp [hunk 1]
```diff
@@ -1394,5 +1394,28 @@ size_t InteractionModelEngine::GetNumDirtySubscriptions() const
...
 
+void InteractionModelEngine::OnFabricRemoved(FabricIndex aFabricIndex)
+{
+    // Try to send the queued events as soon as possible. If the just emitted leave event won't
+    // be sent this time, it will likely not be delivered at all for the following reasons:
+    // - removing the fabric expires all associated ReadHandlers, so all subscriptions to
+    //   the leave event will be cancelled.
+    // - removing the fabric removes all associated access control entries, so generating
+    //   subsequent reports containing the leave event will fail the access control check.
+    GetReportingEngine().ScheduleUrgentEventDeliverySync();
```
Comment (Damian-Nordic) [inline src/app/InteractionModelEngine.cpp:1405]: Be aware of https://github.com/project-chip/connectedhomeip/pull/21191/files. This must be called before ACLs are removed.

Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:1405]: The Leave event already got sent, in `FabricWillBeRemoved`.  So you don't need to do anything with it here.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Tthis needs a merge/rebase to tip.

File: src/app/InteractionModelEngine.cpp (hunk 1 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:1405]: updated. thanks

Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:1405]: updated, thanks

File: src/app/InteractionModelEngine.cpp [hunk 2]
```diff
@@ -1408,5 +1408,32 @@ size_t InteractionModelEngine::GetNumDirtySubscriptions() const
     return numDirtySubscriptions;
 }
 
+void InteractionModelEngine::OnFabricRemoved(FabricIndex aFabricIndex)
```
Comment (mrjerryjohns) [inline src/app/InteractionModelEngine.cpp:1411]: Where is similar logic for shutting down active `ReadClient`s?

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 3]
```diff
@@ -378,6 +378,8 @@ class OpCredsFabricTableDelegate : public chip::FabricTable::Delegate
         ChipLogProgress(Zcl, "OpCreds: Fabric index 0x%x was removed", static_cast<unsigned>(fabricIndex));
 
         EventManagement::GetInstance().FabricRemoved(fabricIndex);
+        InteractionModelEngine::GetInstance()->OnFabricRemoved(fabricIndex);
```
Comment (mrjerryjohns) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:381]: Why is this here? Why isn't the IM engine directly registering itself as a FabricTable::Delegate by calling `mpFabricTable->AddFabricDelegate` at init time?

That will ensure it works consistently across both server and controller modalities.

File: src/app/tests/TestReadInteraction.cpp [hunk 4]
```diff
@@ -1855,6 +1856,64 @@ void TestReadInteraction::TestSubscribeWildcard(nlTestSuite * apSuite, void * ap
...
+
+        // We have 29 attributes in our mock attribute storage. And we subscribed twice.
+        // And attribute 3/2/4 is a list with 6 elements and list chunking is applied to it, thus we should receive ( 29 + 6 ) * 2 =
+        // 70 attribute data in total.
+        NL_TEST_ASSERT(apSuite, delegate.mNumAttributeResponse == 70);
+        NL_TEST_ASSERT(apSuite, engine->GetNumActiveReadHandlers(ReadHandler::InteractionType::Subscribe) == 1);
+        NL_TEST_ASSERT(apSuite, engine->ActiveHandlerAt(0) != nullptr);
+        delegate.mpReadHandler = engine->ActiveHandlerAt(0);
+        InteractionModelEngine::GetInstance()->OnFabricRemoved(delegate.mpReadHandler->GetAccessingFabricIndex());
+        NL_TEST_ASSERT(apSuite, engine->GetNumActiveReadHandlers(ReadHandler::InteractionType::Subscribe) == 0);
```
Comment (mrjerryjohns) [inline src/app/tests/TestReadInteraction.cpp:1909]: This isn't quite a good test, since it's just validating that this method returns correctly, which doesn't really mean subscriptions have been terminated.

The right kind of test is to actually confirm on the subscriber side that the sub has gone away.

To that end, I've just recently added a test to validate resubscriptions in `TestRead.cpp`. Check that out. Using similar logic, you can check that a re-subscription got triggered, and use that to prove the disappearance of sub on the server side.

File: src/controller/tests/TestWriteChunking.cpp [hunk 5]
```diff
@@ -255,6 +256,52 @@ void TestWriteChunking::TestListChunking(nlTestSuite * apSuite, void * apContext
...
+    ctx.GetLoopback().mSentMessageCount            = 0;
+    ctx.GetLoopback().mNumMessagesToDrop           = 1;
+    ctx.GetLoopback().mNumMessagesToDropSinceIndex = 3;
+    err                                            = writeClient.SendWriteRequest(sessionHandle);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    ctx.DrainAndServiceIO();
+
+    NL_TEST_ASSERT(apSuite, InteractionModelEngine::GetInstance()->GetNumActiveWriteHandlers() == 1);
+    InteractionModelEngine::GetInstance()->OnFabricRemoved(1);
+    NL_TEST_ASSERT(apSuite, InteractionModelEngine::GetInstance()->GetNumActiveWriteHandlers() == 0);
```
Comment (mrjerryjohns) [inline src/controller/tests/TestWriteChunking.cpp:297]: Similar to comment above, need to validate on client side.

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

File: src/app/InteractionModelEngine.cpp [hunk 6]
```diff
@@ -1408,5 +1408,32 @@ size_t InteractionModelEngine::GetNumDirtySubscriptions() const
...
+        {
+            ChipLogProgress(InteractionModel, "Fabric removed, deleting obsolete write handler with FabricIndex: %u",
+                            handler.GetAccessingFabricIndex());
+            handler.Close();
+        }
+    }
+
+    // App code may hold the reference for Async Command Handler, when specific fabric index is removed, if we are not cleaning up
+    // the corresponding CommandHandler holed by the app, the app may fail to send Command Response using invalid exchange, then it
+    // is still able to close
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:1436]:
```suggestion
    // Applications may hold references to CommandHandler instances for async command processing.
    // Therefore we can't forcible destroy CommandHandlers here.  Their exchanges will get closed by
    // the fabric removal, though, so they will fail when they try to actually send their command response
    // and will close at that point.
```
As a followup, maybe CommandHandler should track its handles so it's able to release them all or something....  We'd need to check consumers to make sure they do the requisite null-checks.

File: src/app/WriteHandler.h [hunk 7]
```diff
@@ -76,6 +76,11 @@ class WriteHandler : public Messaging::ExchangeDelegate
      */
     void Abort();
```
Comment (bzbarsky-apple) [inline src/app/WriteHandler.h:77]: A followup: we should just remove `Abort`.  It's only used from `InteractionModelEngine::Shutdown`, which could just as easily check `IsFree` and call `Close`.

File: src/transport/raw/tests/NetworkTestHelpers.h [hunk 8]
```diff
@@ -64,7 +64,7 @@ class LoopbackTransportDelegate
 public:
     virtual ~LoopbackTransportDelegate() {}
 
-    // Called by the loopback transport when it drops a message due to a nonzero mNumMessagesToDrop.
+    // Called by the loopback transport when it drops a message due to a nonzero mNumMessagesToDrop/mNumMessagesToDropSinceIndex
```
Comment (bzbarsky-apple) [inline src/transport/raw/tests/NetworkTestHelpers.h:67]: Please rebase on top of https://github.com/project-chip/connectedhomeip/pull/21321

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/InteractionModelEngine.cpp (hunk 2 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp]: added, thanks

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp]: added, thanks

File: src/app/tests/TestReadInteraction.cpp (hunk 4 above)
Comment (yunhanw-google) [reply src/app/tests/TestReadInteraction.cpp]: per offline chat, use GetFabricTable().Delete to remove fabric in test.

File: src/controller/tests/TestWriteChunking.cpp (hunk 5 above)
Comment (yunhanw-google) [reply src/controller/tests/TestWriteChunking.cpp]: per offline chat, use GetFabricTable().Delete to remove fabric in test.

File: src/app/InteractionModelEngine.cpp [hunk 9]
```diff
@@ -1408,5 +1408,32 @@ size_t InteractionModelEngine::GetNumDirtySubscriptions() const
...
 }
 
+void InteractionModelEngine::OnFabricRemoved(FabricIndex aFabricIndex)
+{
+    mReadHandlers.ForEachActiveObject([this, aFabricIndex](ReadHandler * handler) {
+        if (handler->GetAccessingFabricIndex() == aFabricIndex)
+        {
+            ChipLogProgress(InteractionModel, "Fabric removed, deleting obsolete read/subscription handler with FabricIndex: %u",
+                            handler->GetAccessingFabricIndex());
+            mReadHandlers.ReleaseObject(handler);
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp]: agree, updated with Close on the handler.

File: src/transport/raw/tests/NetworkTestHelpers.h (hunk 8 above)
Comment (yunhanw-google) [reply src/transport/raw/tests/NetworkTestHelpers.h]: rebased, thanks

File: src/app/WriteHandler.h (hunk 7 above)
Comment (yunhanw-google) [reply src/app/WriteHandler.h:77]: https://github.com/project-chip/connectedhomeip/issues/21739

File: src/app/InteractionModelEngine.cpp (hunk 6 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:1450]: https://github.com/project-chip/connectedhomeip/issues/21739

File: src/app/WriteHandler.h (hunk 7 above)
Comment (yunhanw-google) [reply src/app/WriteHandler.h:77]: https://github.com/project-chip/connectedhomeip/issues/21740

File: src/app/InteractionModelEngine.cpp [hunk 10]
```diff
@@ -1434,5 +1435,41 @@ size_t InteractionModelEngine::GetNumDirtySubscriptions() const
...
+
+        return Loop::Continue;
+    });
+
+    for (auto * readClient = mpActiveReadClientList; readClient != nullptr; readClient = readClient->GetNextClient())
+    {
+        if (readClient->GetFabricIndex() == fabricIndex)
+        {
+            ChipLogProgress(InteractionModel, "Fabric removed, deleting obsolete read client with FabricIndex: %u", fabricIndex);
+            readClient->Close(CHIP_NO_ERROR, false);
```
Comment (mrjerryjohns) [inline src/app/InteractionModelEngine.cpp:1455]: I think a more apt error code conveying fabric deletion would be better here.

File: src/app/InteractionModelEngine.cpp [hunk 11]
```diff
@@ -1434,5 +1435,41 @@ size_t InteractionModelEngine::GetNumDirtySubscriptions() const
...
+            RemoveReadClient(readClient);
```
Comment (mrjerryjohns) [inline src/app/InteractionModelEngine.cpp:1456]: Why not set `readClient.mpEngine` to null like we do at shutdown? That will avoid the need to check `InActiveReadClientList` later in ReadClient's destructor.

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 12]
```diff
@@ -379,11 +378,16 @@ class OpCredsFabricTableDelegate : public chip::FabricTable::Delegate
...
         EventManagement::GetInstance().FabricRemoved(fabricIndex);
+
         NotifyFabricTableChanged();
     }
 
     // Gets called when a fabric is added/updated, but not necessarily committed to storage
-    void OnFabricUpdated(const FabricTable & fabricTable, FabricIndex fabricIndex) override { NotifyFabricTableChanged(); }
+    void OnFabricUpdated(const FabricTable & fabricTable, FabricIndex fabricIndex) override
+    {
+        ChipLogProgress(InteractionModel, "OnFabricUpdated debugging from opertonal-credential!!!!!!!");
```
Comment (mrjerryjohns) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:388]: Left-over debug prints? :)

File: src/app/InteractionModelEngine.cpp (hunk 11 above)
Comment (mrjerryjohns) [reply src/app/InteractionModelEngine.cpp:1456]: You'll need to reverse the order - you have to remove the client first, then call Close on it.

File: src/app/tests/TestReadInteraction.cpp [hunk 13]
```diff
@@ -3651,6 +3652,71 @@ void TestReadInteraction::TestReadHandlerInvalidSubscribeRequest(nlTestSuite * a
...
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        ctx.DrainAndServiceIO();
+
+        NL_TEST_ASSERT(apSuite, delegate.mGotReport);
+
+        // We have 29 attributes in our mock attribute storage. And we subscribed twice.
+        // And attribute 3/2/4 is a list with 6 elements and list chunking is applied to it, thus we should receive ( 29 + 6 ) * 2 =
+        // 70 attribute data in total.
+        NL_TEST_ASSERT(apSuite, delegate.mNumAttributeResponse == 70);
```
Comment (mrjerryjohns) [inline src/app/tests/TestReadInteraction.cpp:3699]: I'm not a big fan of these still asserts that validate a specific number of attributes received, since it's quite brittle and can break if an attribute was added/removed.

We can just subscribe to a single attribute right?

File: src/app/tests/TestReadInteraction.cpp [hunk 14]
```diff
@@ -3651,6 +3652,71 @@ void TestReadInteraction::TestReadHandlerInvalidSubscribeRequest(nlTestSuite * a
...
+        NL_TEST_ASSERT(apSuite, engine->GetNumActiveReadHandlers(ReadHandler::InteractionType::Subscribe) == 1);
+        NL_TEST_ASSERT(apSuite, engine->ActiveHandlerAt(0) != nullptr);
+        delegate.mpReadHandler = engine->ActiveHandlerAt(0);
+
+        ctx.GetFabricTable().Delete(ctx.GetAliceFabricIndex());
+        NL_TEST_ASSERT(apSuite, engine->GetNumActiveReadHandlers(ReadHandler::InteractionType::Subscribe) == 0);
+        ctx.GetFabricTable().Delete(ctx.GetBobFabricIndex());
+        NL_TEST_ASSERT(apSuite, engine->GetNumActiveReadClients() == 0);
+        ctx.ExpireSessionAliceToBob();
```
Comment (mrjerryjohns) [inline src/app/tests/TestReadInteraction.cpp:3708]: Why is this needed?

Comment (mrjerryjohns) [reply src/app/tests/TestReadInteraction.cpp:3708]: Wouldn't the act of removing the fabric have automatically deleted the session too?

File: src/app/tests/TestWriteInteraction.cpp [hunk 15]
```diff
@@ -626,6 +627,56 @@ void TestWriteInteraction::TestWriteHandlerReceiveInvalidMessage(nlTestSuite * a
...
+    auto * engine = chip::app::InteractionModelEngine::GetInstance();
+    err           = engine->Init(&ctx.GetExchangeManager(), &ctx.GetFabricTable());
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    app::WriteClient writeClient(&ctx.GetExchangeManager(), &writeCallback, Optional<uint16_t>::Missing(),
+                                 static_cast<uint16_t>(900) /* reserved buffer size */);
+
+    ByteSpan list[5];
+
+    err = writeClient.EncodeAttribute(attributePath, app::DataModel::List<ByteSpan>(list, 5));
```
Comment (mrjerryjohns) [inline src/app/tests/TestWriteInteraction.cpp:655]: Aren't you encoding data from a span that's currently pointing to ... garbage?

File: src/app/InteractionModelEngine.cpp [hunk 16]
```diff
@@ -76,9 +77,9 @@ void InteractionModelEngine::Shutdown()
     //
     while (handlerIter)
     {
-        CommandHandlerInterface * next = handlerIter->GetNext();
+        CommandHandlerInterface * pNext = handlerIter->GetNext();
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:80]: Given that the declaration is right there next to all the uses of this variable, I'm not sure the `p` prefix is needed... Same for some of the other cases like that in this PR.  And it does make the readability worse...

File: src/app/InteractionModelEngine.cpp (hunk 11 above)
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:1456]: Why do we need to do this manual removal at all?

File: src/app/InteractionModelEngine.h [hunk 17]
```diff
@@ -289,6 +290,8 @@ class InteractionModelEngine : public Messaging::UnsolicitedMessageHandler,
      */
     uint16_t GetMinGuaranteedSubscriptionsPerFabric() const;
 
+    void OnFabricRemoved(const FabricTable & fabricTable, FabricIndex fabricIndex) override;
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.h:293]: Please document this is coming from FabricTable::Delegate.

File: src/app/ReadClient.cpp [hunk 18]
```diff
@@ -86,7 +86,7 @@ ReadClient::~ReadClient()
         // This won't be the case if the engine shut down before this destructor was called (in which case, mpImEngine
         // will point to null)
         //
-        if (mpImEngine)
+        if (mpImEngine && mpImEngine->InActiveReadClientList(this))
```
Comment (bzbarsky-apple) [inline src/app/ReadClient.cpp:89]: Why do we need this check?  If we remove this check _and_ remove the manual removal above, won't things work correctly?

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 19]
```diff
@@ -282,7 +282,6 @@ CHIP_ERROR DeleteFabricFromTable(FabricIndex fabricIndex)
 
 void CleanupSessionsForFabric(SessionManager & sessionMgr, FabricIndex fabricIndex)
 {
-    InteractionModelEngine::GetInstance()->CloseTransactionsFromFabricIndex(fabricIndex);
```
Comment (bzbarsky-apple) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:285]: So what calls CloseTransactionsFromFabricIndex now?  I don't see it being called from the delegate callback in IM engine.  And if the delegate callback is doing its work now, should CloseTransactionsFromFabricIndex just go away?

File: src/app/tests/TestWriteInteraction.cpp (hunk 15 above)
Comment (bzbarsky-apple) [reply src/app/tests/TestWriteInteraction.cpp:655]: It's pointing to default-constructed Spans.  Span default constructs to (nullptr, 0), aka empty span.

File: src/app/InteractionModelEngine.cpp (hunk 10 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp]: Done, thanks

File: src/app/InteractionModelEngine.cpp (hunk 11 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:1455]: Done, thanks

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp (hunk 12 above)
Comment (yunhanw-google) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp]: removed. thanks

File: src/app/tests/TestWriteInteraction.cpp (hunk 15 above)
Comment (yunhanw-google) [reply src/app/tests/TestWriteInteraction.cpp:655]: yes

File: src/app/InteractionModelEngine.cpp (hunk 16 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:80]: hip::app::InteractionModelEngine::Shutdown()’:
../../src/app/InteractionModelEngine.cpp:80:35: error: declaration of ‘next’ shadows a member of ‘chip::app::InteractionModelEngine’ [-Werror=shadow]
   80 |         CommandHandlerInterface * next = handlerIter->GetNext();
      |                                   ^~~~
In file included from ../../src/transport/SessionManager.h:30,
                 from ../../src/messaging/ExchangeMessageDispatch.h:28,
                 from ../../src/messaging/ApplicationExchangeDispatch.h:28,
                 from ../../src/messaging/ExchangeDelegate.h:27,
                 from ../../src/messaging/ExchangeContext.h:31,
                 from ../../src/app/InteractionModelEngine.h:36,
                 from ../../src/app/InteractionModelEngine.cpp:26:
../../src/credentials/FabricTable.h:388:20: note: shadowed declaration is here
  388 |         Delegate * next = nullptr;
      |                    ^~~~
At global scope:
cc1plus: note: unrecognized command-line option ‘-Wno-unknown-warning-option’ may have been intended to silence earlier diagnostics
cc1plus: all warnings being treated as errors
[2000/4192] ACTION //src/controller/tests:TestWriteChunking_run(//build/toolchain/host:linux_x64_clang)
ninja: build stopped: subcommand failed.
real	1m15.373s

Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:80]: renamed it to nextHandler

File: src/app/InteractionModelEngine.cpp (hunk 11 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:1455]: since the pointer removal is happening inside read client destructor.

ReadClient::~ReadClient()
{
    if (IsSubscriptionType())
    {
        CancelLivenessCheckTimer();
        CancelResubscribeTimer();

        //
        // Only remove ourselves from the engine's tracker list if we still continue to have a valid pointer to it.
        // This won't be the case if the engine shut down before this destructor was called (in which case, mpImEngine
        // will point to null)
        //
        if (mpImEngine)
        {
            mpImEngine->RemoveReadClient(this);
        }
    }
}

File: src/app/InteractionModelEngine.h (hunk 17 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:293]: since the pointer removal is happening inside read client destructor.

Comment (yunhanw-google) [reply src/app/InteractionModelEngine.h:293]: done, thanks

File: src/app/ReadClient.cpp (hunk 18 above)
Comment (yunhanw-google) [reply src/app/ReadClient.cpp]: since the pointer removal is happening inside read client destructor.

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp (hunk 19 above)
Comment (yunhanw-google) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:285]: https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/operational-credentials-server/operational-credentials-server.cpp#L325
https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/operational-credentials-server/operational-credentials-server.cpp#L313

Based upon the above call flow, 
yes, i think CloseTransactionsFromFabricIndex method should be removed.

File: src/app/tests/TestReadInteraction.cpp (hunk 14 above)
Comment (yunhanw-google) [reply src/app/tests/TestReadInteraction.cpp:3708]: I think FabricTable.Delete itself don't expire session. 

If you operational credential code, in failsafe cleanup, session release first, fabric deletion second.

https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/operational-credentials-server/operational-credentials-server.cpp#L308

https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/operational-credentials-server/operational-credentials-server.cpp#L325

File: src/app/InteractionModelEngine.cpp (hunk 11 above)
Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp:1438]: > since the pointer removal is happening inside read client destructor.

OK, but so what?  What's special about this Close() call compared to other Close() calls that don't do the manual removal?

File: src/app/tests/TestReadInteraction.cpp (hunk 13 above)
Comment (yunhanw-google) [reply src/app/tests/TestReadInteraction.cpp]: yes, resolved, thanks

File: src/app/InteractionModelEngine.cpp (hunk 11 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp]: per chat with boris, we should do the same thing here no matter whether the Close is coming from this code or from an error on the wire: we should either remove in both cases or not remove in both cases. 
I just remove the manual removal

File: src/app/ReadClient.cpp (hunk 18 above)
Comment (yunhanw-google) [reply src/app/ReadClient.cpp]: just remove the manual removal

File: src/app/InteractionModelEngine.cpp [hunk 20]
```diff
@@ -1217,9 +1200,9 @@ void InteractionModelEngine::ReleasePool(ObjectList<T> *& aObjectList, ObjectPoo
     ObjectList<T> * current = aObjectList;
     while (current != nullptr)
     {
-        ObjectList<T> * next = current->mpNext;
+        ObjectList<T> * pNext = current->mpNext;
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:1203]: nextObject

File: src/lib/core/CHIPError.h [hunk 21]
```diff
@@ -2428,6 +2428,14 @@ using CHIP_ERROR = ::chip::ChipError;
  */
 #define CHIP_ERROR_MISSING_URI_SEPARATOR                      CHIP_CORE_ERROR(0xe0)
 
+/**
+ * @def CHIP_ERROR_IM_FABRIC_DELETED
+ *
+ * @brief
+ *   The fabric is deleted, and the corresponding IM resources are released
+ */
+#define CHIP_ERROR_IM_FABRIC_DELETED                            CHIP_CORE_ERROR(0xe1)
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPError.h:2437]: Please use on of the existing skipped values earlier in this file.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/InteractionModelEngine.cpp (hunk 20 above)
Comment (yunhanw-google) [reply src/app/InteractionModelEngine.cpp:1203]: Done, thanks

File: src/lib/core/CHIPError.h (hunk 21 above)
Comment (yunhanw-google) [reply src/lib/core/CHIPError.h:2437]: Done, just replace the unused one. thanks

**Review by mrjerryjohns — APPROVED** [review]
